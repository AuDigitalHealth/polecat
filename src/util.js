// SHA-256 function from
// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
export async function sha256(str) {
  // We transform the string into an arraybuffer.
  const buffer = str2ab(str)
  return crypto.subtle.digest('SHA-256', buffer).then(function(hash) {
    return hex(hash)
  })
}

function str2ab(str) {
  const buf = new ArrayBuffer(str.length * 2) // 2 bytes for each char
  const bufView = new Uint16Array(buf)
  for (let i = 0, strLen = str.length; i < strLen; i++) {
    bufView[i] = str.charCodeAt(i)
  }
  return buf
}

function hex(buffer) {
  const hexCodes = []
  const view = new DataView(buffer)
  for (let i = 0; i < view.byteLength; i += 4) {
    // Using getUint32 reduces the number of iterations needed (we process 4
    // bytes each time)
    const value = view.getUint32(i)
    // toString(16) will give the hex representation of the number without
    // padding
    const stringValue = value.toString(16)
    // We use concatenation and slice for padding
    const padding = '00000000'
    const paddedValue = (padding + stringValue).slice(-padding.length)
    hexCodes.push(paddedValue)
  }

  // Join all the hex strings into one
  return hexCodes.join('')
}

export const formatNumber = number => {
  if (typeof number !== 'number') return undefined
  const numString = number.toString()
  return [...numString].reduce((prev, curr, i, arr) => {
    return arr.length - i - 1 !== 0 && (arr.length - i - 1) % 3 === 0
      ? prev + curr + ','
      : prev + curr
  }, '')
}

export const parseUri = str => {
  var o = parseUri.options,
    m = o.parser[o.strictMode ? 'strict' : 'loose'].exec(str),
    uri = {},
    i = 14

  while (i--) uri[o.key[i]] = m[i] || ''

  uri[o.q.name] = {}
  uri[o.key[12]].replace(o.q.parser, function($0, $1, $2) {
    if ($1) uri[o.q.name][$1] = $2
  })

  return uri
}

parseUri.options = {
  strictMode: false,
  key: [
    'source',
    'protocol',
    'authority',
    'userInfo',
    'user',
    'password',
    'host',
    'port',
    'relative',
    'path',
    'directory',
    'file',
    'query',
    'anchor',
  ],
  q: {
    name: 'queryKey',
    parser: /(?:^|&)([^&=]*)=?([^&]*)/g,
  },
  parser: {
    // eslint-disable-next-line no-useless-escape
    strict: /^(?:([^:\/?#]+):)?(?:\/\/((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?))?((((?:[^?#\/]*\/)*)([^?#]*))(?:\?([^#]*))?(?:#(.*))?)/,
    // eslint-disable-next-line no-useless-escape
    loose: /^(?:(?![^:@]+:[^:@\/]*@)([^:\/?#.]+):)?(?:\/\/)?((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?)(((\/(?:[^?#](?![^?#\/]*\.[^?#\/.]+(?:[?#]|$)))*\/?)?([^?#\/]*))(?:\?([^#]*))?(?:#(.*))?)/,
  },
}
