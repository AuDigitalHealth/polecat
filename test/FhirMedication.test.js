import React from 'react'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import crypto from 'crypto'

import FhirMedication from '../src/FhirMedication.js'
import * as util from '../src/util.js'

import resource1 from './fixtures/mpuu-23628011000036109.json'
import resource1ChildBundle1 from './fixtures/childBundle-mpuu-23628011000036109.json'
import resource1ChildBundle2 from './fixtures/childBundle-tpuu-23628011000036109.json'
import resource1PackageBundle from './fixtures/packageBundle-mpp-23628011000036109.json'

import resource2 from './fixtures/substance-2292011000036106.json'
import resource2IngredientBundle1 from './fixtures/ingredientBundle-mp-2292011000036106.json'
import resource2IngredientBundle2 from './fixtures/ingredientBundle-mpuu-2292011000036106.json'
import resource2IngredientBundle3 from './fixtures/ingredientBundle-tpuu-2292011000036106.json'

import resource3 from './fixtures/mp-21360011000036101.json'

import resource4 from './fixtures/tpuu-6052011000036107.json'

describe('FhirMedication', () => {
  beforeEach(() => {
    // Replace the browser-dependent `sha256` function with a Node-compatible implementation.
    sinon.stub(util, 'sha256').callsFake(str => {
      const hash = crypto.createHash('sha256')
      hash.update(str)
      return Promise.resolve(hash.digest('hex'))
    })
  })

  afterEach(() => util.sha256.restore())

  it('should pass nodes and links through for UPDSF', () => {
    const props = {
      resource: resource1,
      childBundles: {
        UPDSF: resource1ChildBundle1,
        BPSF: resource1ChildBundle2,
      },
      packageBundles: { UPG: resource1PackageBundle },
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    // Give the async code a chance to execute, then check that the correct
    // nodes and links have been passed through to the child component.
    return new Promise(resolve => {
      setTimeout(() => {
        wrapper.update()
        const nodes = wrapper.prop('nodes'),
          links = wrapper.prop('links')
        expect(nodes).toMatchSnapshot()
        expect(links).toMatchSnapshot()
        resolve()
      }, 50)
    })
  })

  it('should pass nodes and links through for substance', () => {
    const props = {
      resource: resource2,
      containsIngredientBundles: {
        UPD: resource2IngredientBundle1,
        UPDSF: resource2IngredientBundle2,
        BPSF: resource2IngredientBundle3,
      },
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    // Give the async code a chance to execute, then check that the correct
    // nodes and links have been passed through to the child component.
    return new Promise(resolve => {
      setTimeout(() => {
        wrapper.update()
        const nodes = wrapper.prop('nodes'),
          links = wrapper.prop('links')
        expect(nodes).toMatchSnapshot()
        expect(links).toMatchSnapshot()
        resolve()
      }, 50)
    })
  })

  it('should update the primary resource', () => {
    const props = {
      resource: resource1,
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    // Update the resource prop.
    wrapper.setProps({ resource: resource2 })
    // Give the async code a chance to execute, then check that the correct
    // nodes and links have been passed through to the child component.
    return new Promise(resolve => {
      setTimeout(() => {
        wrapper.update()
        const nodes = wrapper.prop('nodes')
        expect(nodes).toMatchSnapshot()
        resolve()
      }, 50)
    })
  })

  it('should update child bundles', () => {
    const props = {
      resource: resource1,
      childBundles: {
        UPDSF: resource1ChildBundle1,
      },
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    wrapper.update()
    return new Promise(resolve => {
      setTimeout(() => {
        // Give the mounted state a chance to resolve itself, then push through
        // a change to the `childBundles` prop.
        wrapper.setProps({
          childBundles: {
            UPDSF: resource1ChildBundle1,
            BPSF: resource1ChildBundle2,
          },
        })
        setTimeout(() => {
          // Once the change to the prop has finished, check that the right
          // value has flowed through to the `nodes` prop on the child component.
          wrapper.update()
          const nodes = wrapper.prop('nodes')
          expect(nodes).toMatchSnapshot()
          resolve()
        }, 50)
      }, 50)
    })
  })

  it('should request child UPDs for subject UPD', () => {
    const onRequireChildBundle = jest.fn()
    const props = {
      resource: resource3,
      onRequireChildBundle,
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    // Give the async code a chance to execute, then check that the callback has
    // been called with the expected arguments.
    return new Promise(resolve => {
      setTimeout(() => {
        wrapper.update()
        expect(onRequireChildBundle).toHaveBeenCalledWith(
          '21360011000036101',
          'UPD',
        )
        resolve()
      }, 50)
    })
  })

  it('should request BPG packages for subject BPSF', () => {
    const onRequirePackageBundle = jest.fn()
    const props = {
      resource: resource4,
      onRequirePackageBundle,
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    // Give the async code a chance to execute, then check that the callback has
    // been called with the expected arguments.
    return new Promise(resolve => {
      setTimeout(() => {
        wrapper.update()
        expect(onRequirePackageBundle).toHaveBeenCalledWith(
          '6052011000036107',
          'BPG',
        )
        resolve()
      }, 50)
    })
  })

  it('should request UPDs containing subject substance', () => {
    const onRequireContainsIngredientBundle = jest.fn()
    const props = {
      resource: resource2,
      onRequireContainsIngredientBundle,
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    // Give the async code a chance to execute, then check that the callback has
    // been called with the expected arguments.
    return new Promise(resolve => {
      setTimeout(() => {
        wrapper.update()
        expect(onRequireContainsIngredientBundle).toHaveBeenCalledWith(
          '2292011000036106',
          'UPD',
        )
        resolve()
      }, 50)
    })
  })

  it('should notify about load of subject concept', () => {
    const onLoadSubjectConcept = jest.fn()
    const props = {
      resource: resource1,
      onLoadSubjectConcept,
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    // Give the async code a chance to execute, then check that the callback has
    // been called with the expected arguments.
    return new Promise(resolve => {
      setTimeout(() => {
        wrapper.update()
        expect(onLoadSubjectConcept).toHaveBeenCalledWith({
          coding: [
            {
              code: '23628011000036109',
              display: 'paracetamol 500 mg tablet',
              system: 'http://snomed.info/sct',
            },
          ],
          sourceCodeSystemUri: 'http://snomed.info/sct',
          sourceCodeSystemVersion:
            'http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228',
          lastModified: '2014-06-30',
          status: 'active',
          type: 'UPDSF',
        })
        resolve()
      }, 50)
    })
  })

  it('should not request package bundles twice when updated with child bundles', () => {
    const onRequirePackageBundle = jest.fn()
    const props = {
      resource: resource1,
      childBundles: {
        UPDSF: resource1ChildBundle1,
      },
      onRequirePackageBundle,
    }
    const wrapper = shallow(
      <FhirMedication {...props}>
        <div />
      </FhirMedication>,
    )
    wrapper.update()
    return new Promise(resolve => {
      setTimeout(() => {
        // Check that the `onRequirePackageBundle` callback has been called
        // once, at this point.
        expect(onRequirePackageBundle).toHaveBeenCalledTimes(1)
        onRequirePackageBundle.mockClear()
        // Give the mounted state a chance to resolve itself, then push through
        // a change to the `childBundles` prop.
        wrapper.setProps({
          childBundles: {
            UPDSF: resource1ChildBundle1,
            BPSF: resource1ChildBundle2,
          },
        })
        setTimeout(() => {
          // Once the change to the prop has finished, check that the
          // `onRequirePackageBundle` callback has not been called again.
          wrapper.update()
          expect(onRequirePackageBundle).not.toHaveBeenCalled()
          resolve()
        }, 50)
      }, 50)
    })
  })
})
