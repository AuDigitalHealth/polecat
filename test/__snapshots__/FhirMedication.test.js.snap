// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FhirMedication should pass nodes and links through for UPDSF 1`] = `
Array [
  Object {
    "coding": Array [
      Object {
        "code": "21433011000036107",
        "display": "paracetamol",
        "system": "http://snomed.info/sct",
      },
    ],
    "status": "active",
    "type": "UPD",
  },
  Object {
    "coding": Array [
      Object {
        "code": "2442011000036104",
        "display": "paracetamol",
        "system": "http://snomed.info/sct",
      },
    ],
    "type": "substance",
  },
  Object {
    "coding": Array [
      Object {
        "code": "23628011000036109",
        "display": "paracetamol 500 mg tablet",
        "system": "http://snomed.info/sct",
      },
    ],
    "focused": true,
    "lastModified": "2014-06-30",
    "sourceCodeSystemUri": "http://snomed.info/sct",
    "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
    "status": "active",
    "subsidy": Array [],
    "type": "UPDSF",
  },
  Object {
    "coding": Array [
      Object {
        "code": "af136f0607369d4c0fba98771e8b05d13237fe381319b59e6a375520615a09a8",
        "system": "group",
      },
    ],
    "concepts": Array [
      Object {
        "coding": Array [
          Object {
            "code": "52258011000036105",
            "display": "paracetamol 500 mg + codeine phosphate hemihydrate 10 mg tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "52806011000036108",
            "display": "paracetamol 500 mg + codeine phosphate hemihydrate 10 mg + doxylamine succinate 5.1 mg tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-02-29",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "52803011000036102",
            "display": "paracetamol 500 mg + codeine phosphate hemihydrate 15 mg tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
    ],
    "query": "type:MPUU ancestor:\\"23628011000036109|paracetamol 500 mg tablet\\"",
    "total": 28,
    "type": "group",
  },
  Object {
    "coding": Array [
      Object {
        "code": "bd1d58572d7bfde3ec89b383060ae08f73a116d33ab055ebb4eeef82b6e5c2c9",
        "system": "group",
      },
    ],
    "concepts": Array [
      Object {
        "coding": Array [
          Object {
            "code": "54282011000036105",
            "display": "Paracetamol (Herron) 500 mg uncoated tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-08-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "88341000036103",
            "display": "Calmative Pain Relief (Pharmacy Action) uncoated tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-08-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "1062931000168103",
            "display": "Pain Relief Paracetamol plus Codeine with Calmative (Cipla) uncoated tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2018-01-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
    ],
    "query": "type:TPUU ancestor:\\"23628011000036109|paracetamol 500 mg tablet\\"",
    "total": 406,
    "type": "group",
  },
  Object {
    "coding": Array [
      Object {
        "code": "24a723931526a57542783b62c236f912708822e33391a8f48c7de44db669e015",
        "system": "group",
      },
    ],
    "concepts": Array [
      Object {
        "coding": Array [
          Object {
            "code": "63941011000036106",
            "display": "paracetamol 500 mg tablet, 48",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPG",
      },
      Object {
        "coding": Array [
          Object {
            "code": "46260011000036106",
            "display": "paracetamol 500 mg tablet, 24",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPG",
      },
      Object {
        "coding": Array [
          Object {
            "code": "46259011000036105",
            "display": "paracetamol 500 mg tablet, 20",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPG",
      },
    ],
    "query": "type:MPP package:\\"23628011000036109|paracetamol 500 mg tablet\\"",
    "total": 25,
    "type": "group",
  },
]
`;

exports[`FhirMedication should pass nodes and links through for UPDSF 2`] = `
Array [
  Object {
    "source": "23628011000036109",
    "target": "21433011000036107",
    "type": "is-a",
  },
  Object {
    "source": "23628011000036109",
    "target": "2442011000036104",
    "type": "has-boss",
  },
  Object {
    "source": "group-af136f0607369d4c0fba98771e8b05d13237fe381319b59e6a375520615a09a8",
    "sourceIsGroup": true,
    "target": "23628011000036109",
    "type": "is-a",
  },
  Object {
    "source": "group-bd1d58572d7bfde3ec89b383060ae08f73a116d33ab055ebb4eeef82b6e5c2c9",
    "sourceIsGroup": true,
    "target": "23628011000036109",
    "type": "is-a",
  },
  Object {
    "source": "group-24a723931526a57542783b62c236f912708822e33391a8f48c7de44db669e015",
    "sourceIsGroup": true,
    "target": "23628011000036109",
    "type": "has-updsf",
  },
]
`;

exports[`FhirMedication should pass nodes and links through for substance 1`] = `
Array [
  Object {
    "coding": Array [
      Object {
        "code": "2292011000036106",
        "display": "loratadine",
        "system": "http://snomed.info/sct",
      },
    ],
    "focused": true,
    "lastModified": "2014-06-30",
    "status": "active",
    "subsidy": Array [],
    "type": "substance",
  },
  Object {
    "coding": Array [
      Object {
        "code": "21701011000036106",
        "display": "loratadine",
        "system": "http://snomed.info/sct",
      },
    ],
    "lastModified": "2014-06-30",
    "sourceCodeSystemUri": "http://snomed.info/sct",
    "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
    "status": "active",
    "subsidy": Array [],
    "type": "UPD",
  },
  Object {
    "coding": Array [
      Object {
        "code": "44955011000036104",
        "display": "loratadine + pseudoephedrine",
        "system": "http://snomed.info/sct",
      },
    ],
    "lastModified": "2014-06-30",
    "sourceCodeSystemUri": "http://snomed.info/sct",
    "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
    "status": "active",
    "subsidy": Array [],
    "type": "UPD",
  },
  Object {
    "coding": Array [
      Object {
        "code": "b3602d2118e67817b7d7dafa97087afa7fb762392942df56eb95d6fe0a991235",
        "system": "group",
      },
    ],
    "concepts": Array [
      Object {
        "coding": Array [
          Object {
            "code": "45158011000036102",
            "display": "loratadine 5 mg + pseudoephedrine sulfate 120 mg modified release tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-05-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "22427011000036102",
            "display": "loratadine 10 mg tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "86001000036103",
            "display": "loratadine 10 mg effervescent tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
    ],
    "query": "type:MPUU ingredient:2292011000036106|loratadine",
    "total": 7,
    "type": "group",
  },
  Object {
    "coding": Array [
      Object {
        "code": "c5ef4be6b5905c8c5c2c96bf73f687d66865905d93d328abec40a993b31e37d3",
        "system": "group",
      },
    ],
    "concepts": Array [
      Object {
        "coding": Array [
          Object {
            "code": "924539011000036107",
            "display": "Claratyne-D with Decongestant Repetab modified release tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-05-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "54112011000036105",
            "display": "Claratyne 10 mg effervescent tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "649061000168101",
            "display": "Lorastyne (Pharmacy Action) 10 mg uncoated tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-08-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
    ],
    "query": "type:TPUU ingredient:2292011000036106|loratadine",
    "total": 27,
    "type": "group",
  },
]
`;

exports[`FhirMedication should pass nodes and links through for substance 2`] = `
Array [
  Object {
    "source": "21701011000036106",
    "target": "2292011000036106",
    "type": "has-ingredient",
  },
  Object {
    "source": "44955011000036104",
    "target": "2292011000036106",
    "type": "has-ingredient",
  },
  Object {
    "source": "group-b3602d2118e67817b7d7dafa97087afa7fb762392942df56eb95d6fe0a991235",
    "sourceIsGroup": true,
    "target": "2292011000036106",
    "type": "has-boss",
  },
  Object {
    "source": "group-c5ef4be6b5905c8c5c2c96bf73f687d66865905d93d328abec40a993b31e37d3",
    "sourceIsGroup": true,
    "target": "2292011000036106",
    "type": "has-boss",
  },
]
`;

exports[`FhirMedication should update child bundles 1`] = `
Array [
  Object {
    "coding": Array [
      Object {
        "code": "21433011000036107",
        "display": "paracetamol",
        "system": "http://snomed.info/sct",
      },
    ],
    "status": "active",
    "type": "UPD",
  },
  Object {
    "coding": Array [
      Object {
        "code": "2442011000036104",
        "display": "paracetamol",
        "system": "http://snomed.info/sct",
      },
    ],
    "type": "substance",
  },
  Object {
    "coding": Array [
      Object {
        "code": "23628011000036109",
        "display": "paracetamol 500 mg tablet",
        "system": "http://snomed.info/sct",
      },
    ],
    "focused": true,
    "lastModified": "2014-06-30",
    "sourceCodeSystemUri": "http://snomed.info/sct",
    "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
    "status": "active",
    "subsidy": Array [],
    "type": "UPDSF",
  },
  Object {
    "coding": Array [
      Object {
        "code": "af136f0607369d4c0fba98771e8b05d13237fe381319b59e6a375520615a09a8",
        "system": "group",
      },
    ],
    "concepts": Array [
      Object {
        "coding": Array [
          Object {
            "code": "52258011000036105",
            "display": "paracetamol 500 mg + codeine phosphate hemihydrate 10 mg tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "52806011000036108",
            "display": "paracetamol 500 mg + codeine phosphate hemihydrate 10 mg + doxylamine succinate 5.1 mg tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-02-29",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "52803011000036102",
            "display": "paracetamol 500 mg + codeine phosphate hemihydrate 15 mg tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2014-06-30",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "UPDSF",
      },
    ],
    "query": "type:MPUU ancestor:\\"23628011000036109|paracetamol 500 mg tablet\\"",
    "total": 28,
    "type": "group",
  },
  Object {
    "coding": Array [
      Object {
        "code": "bd1d58572d7bfde3ec89b383060ae08f73a116d33ab055ebb4eeef82b6e5c2c9",
        "system": "group",
      },
    ],
    "concepts": Array [
      Object {
        "coding": Array [
          Object {
            "code": "54282011000036105",
            "display": "Paracetamol (Herron) 500 mg uncoated tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-08-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "88341000036103",
            "display": "Calmative Pain Relief (Pharmacy Action) uncoated tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2016-08-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
      Object {
        "coding": Array [
          Object {
            "code": "1062931000168103",
            "display": "Pain Relief Paracetamol plus Codeine with Calmative (Cipla) uncoated tablet",
            "system": "http://snomed.info/sct",
          },
        ],
        "lastModified": "2018-01-31",
        "sourceCodeSystemUri": "http://snomed.info/sct",
        "sourceCodeSystemVersion": "http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228",
        "status": "active",
        "subsidy": Array [],
        "type": "BPSF",
      },
    ],
    "query": "type:TPUU ancestor:\\"23628011000036109|paracetamol 500 mg tablet\\"",
    "total": 406,
    "type": "group",
  },
]
`;

exports[`FhirMedication should update the primary resource 1`] = `
Array [
  Object {
    "coding": Array [
      Object {
        "code": "2292011000036106",
        "display": "loratadine",
        "system": "http://snomed.info/sct",
      },
    ],
    "focused": true,
    "lastModified": "2014-06-30",
    "status": "active",
    "subsidy": Array [],
    "type": "substance",
  },
]
`;
